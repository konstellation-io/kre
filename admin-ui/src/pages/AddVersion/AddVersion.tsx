import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { useMutation } from '@apollo/react-hooks';
import { useHistory, useParams } from 'react-router';

import { get } from 'lodash';

import SpinnerLinear from '../../components/LoadingComponents/SpinnerLinear/SpinnerLinear';
import FileUpload from '../../components/Form/FileUpload/FileUpload';
import Button from '../../components/Button/Button';
import ROUTE, { RuntimeRouteParams } from '../../constants/routes';

import styles from './AddVersion.module.scss';

import { loader } from 'graphql.macro';
import {
  CreateVersion,
  CreateVersionVariables
} from '../../graphql/mutations/types/CreateVersion';
import { buildRoute } from '../../utils/routes';

import { mutationPayloadHelper } from '../../utils/formUtils';

const AddVersionMutation = loader('../../graphql/mutations/addVersion.graphql');

function AddVersion() {
  const history = useHistory();
  const { runtimeId } = useParams<RuntimeRouteParams>();
  const { register, handleSubmit, errors, setError, clearError } = useForm();
  const [addVersion, { loading, error }] = useMutation<
    CreateVersion,
    CreateVersionVariables
  >(AddVersionMutation, {
    onCompleted
  });

  function onCompleted(updatedData: CreateVersion) {
    const versionCreatedId = updatedData.createVersion.id;
    console.log(`${versionCreatedId} version created`);

    history.push(
      buildRoute.version(
        ROUTE.RUNTIME_VERSION_STATUS,
        runtimeId,
        versionCreatedId
      )
    );
  }

  useEffect(() => {
    if (error) {
      setError('addVersionFile', 'apolloError', error.toString());
    } else {
      clearError('addVersionFile');
    }
  }, [error, clearError, setError]);

  function onChange() {
    clearError('addVersionFile');
  }
  function onCancelClick() {
    history.goBack();
  }

  function onSubmit(formData: any) {
    addVersion(
      mutationPayloadHelper({ file: formData.addVersionFile[0], runtimeId })
    );
  }

  return (
    <div className={styles.bg}>
      <div className={styles.grid}>
        <div className={styles.container}>
          <h1>Add Version</h1>
          <p className={styles.subtitle} />
          <div className={styles.content}>
            <form>
              <FileUpload
                name="addVersionFile"
                inputRef={register({
                  required: 'This field is mandatory',
                  validate: value =>
                    (value[0] && value[0].name.includes('.krt')) ||
                    'Must be a .krt file'
                })}
                label="upload version file (KRT file)"
                placeholder=".krt"
                error={get(errors.addVersionFile, 'message', '')}
                onChange={onChange}
              />
              {loading && (
                <div className={styles.spinner}>
                  <SpinnerLinear />
                </div>
              )}
              <div className={styles.buttons}>
                <Button
                  primary
                  disabled={!!errors.addVersionFile || loading}
                  label={'CREATE'}
                  onClick={handleSubmit(onSubmit)}
                />
                <Button label="CANCEL" onClick={onCancelClick} />
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default AddVersion;
