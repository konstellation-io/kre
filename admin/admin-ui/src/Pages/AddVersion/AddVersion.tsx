import { Button, FileUpload, SpinnerLinear } from 'kwc';
import {
  CreateVersion,
  CreateVersionVariables
} from 'Graphql/mutations/types/CreateVersion';
import {
  GetVersionConfStatus,
  GetVersionConfStatus_versions
} from 'Graphql/queries/types/GetVersionConfStatus';
import ROUTE, { RuntimeRouteParams } from 'Constants/routes';
import React, { useEffect, useState } from 'react';
import { useHistory, useParams } from 'react-router';

import { buildRoute } from 'Utils/routes';
import { get } from 'lodash';
import { loader } from 'graphql.macro';
import { mutationPayloadHelper } from 'Utils/formUtils';
import styles from './AddVersion.module.scss';
import { useForm } from 'react-hook-form';
import { useMutation } from '@apollo/client';

const AddVersionMutation = loader('Graphql/mutations/addVersion.graphql');
const GetRuntimeAndVersionQuery = loader(
  'Graphql/queries/getRuntimeAndVersions.graphql'
);

function AddVersion() {
  const history = useHistory();
  const { runtimeId } = useParams<RuntimeRouteParams>();
  const { register, handleSubmit, errors, setError, clearError } = useForm();
  const [addVersion, { loading, error }] = useMutation<
    CreateVersion,
    CreateVersionVariables
  >(AddVersionMutation, {
    onError: e => console.error(`addVersion: ${e}`),
    update(cache, updateResult) {
      if (updateResult.data) {
        const cacheResult = cache.readQuery<GetVersionConfStatus>({
          query: GetRuntimeAndVersionQuery,
          variables: { runtimeId }
        });

        const versions = cacheResult?.versions || [];
        const newVersion = updateResult.data
          .createVersion as GetVersionConfStatus_versions;

        cache.writeQuery({
          query: GetRuntimeAndVersionQuery,
          variables: { runtimeId },
          data: { versions: versions.concat([newVersion]) }
        });
      }
    },
    onCompleted
  });

  const [validationErr, setValidationErr] = useState<string>('');

  function onCompleted(updatedData: CreateVersion) {
    const versionCreatedId = updatedData.createVersion.id;
    console.log(`${versionCreatedId} version created`);

    history.push(buildRoute.runtime(ROUTE.RUNTIME_VERSIONS, runtimeId));
  }

  useEffect(() => {
    if (error) {
      setError('addVersionFile', 'apolloError', error.toString());

      const err = error.graphQLErrors[0];
      if (err.extensions?.code === 'krt_validation_error') {
        setValidationErr(err.extensions?.details);
      }
    } else {
      clearError('addVersionFile');
    }
  }, [error, clearError, setError]);

  function onChange() {
    clearError('addVersionFile');
    setValidationErr('');
  }
  function onCancelClick() {
    history.goBack();
  }

  function onSubmit(formData: any) {
    addVersion(
      mutationPayloadHelper({ file: formData.addVersionFile[0], runtimeId })
    );
  }

  return (
    <div className={styles.bg}>
      <div className={styles.grid}>
        <div className={styles.container}>
          <h1>Add Version</h1>
          <p className={styles.subtitle} />

          {validationErr && (
            <div className={styles.errorBox}>{validationErr}</div>
          )}

          <div className={styles.content}>
            <form>
              <FileUpload
                name="addVersionFile"
                inputRef={register({
                  required: 'This field is mandatory',
                  validate: value =>
                    (value[0] && value[0].name.includes('.krt')) ||
                    'Must be a .krt file'
                })}
                label="upload version file (KRT file)"
                placeholder=".krt"
                error={get(errors.addVersionFile, 'message', '')}
                onChange={onChange}
                autofocus
              />
              {loading && (
                <div className={styles.spinner}>
                  <SpinnerLinear />
                </div>
              )}
              <div className={styles.buttons}>
                <Button
                  primary
                  disabled={!!errors.addVersionFile || loading}
                  label={'CREATE'}
                  onClick={handleSubmit(onSubmit)}
                  className={styles.buttonSave}
                  tabIndex={0}
                />
                <Button label="CANCEL" onClick={onCancelClick} tabIndex={0} />
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default AddVersion;
