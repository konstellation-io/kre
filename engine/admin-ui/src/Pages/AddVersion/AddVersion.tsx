import { Button, FileUpload, SpinnerLinear } from 'kwc';
import {
  CreateVersion,
  CreateVersionVariables
} from 'Graphql/mutations/types/CreateVersion';
import {
  GetVersionConfStatus,
  GetVersionConfStatus_versions
} from 'Graphql/queries/types/GetVersionConfStatus';
import React, { useEffect, useState } from 'react';

import ROUTE, {RuntimeRouteParams} from 'Constants/routes';
import { get } from 'lodash';
import { mutationPayloadHelper } from 'Utils/formUtils';
import styles from './AddVersion.module.scss';
import { useForm } from 'react-hook-form';
import { useHistory } from 'react-router';
import { useMutation } from '@apollo/client';

import AddVersionMutation from 'Graphql/mutations/addVersion';
import GetRuntimeAndVersionQuery from 'Graphql/queries/getRuntimeAndVersions';
import { useParams } from "react-router-dom";
import { buildRoute } from "Utils/routes";

function AddVersion() {
  const history = useHistory();
  const { runtimeId } = useParams<RuntimeRouteParams>();
  const { register, handleSubmit, errors, setError, clearError } = useForm();
  const [addVersion, { loading, error }] = useMutation<
    CreateVersion,
    CreateVersionVariables
  >(AddVersionMutation, {
    onError: e => console.error(`addVersion: ${e}`),
    update(cache, updateResult) {
      if (updateResult.data) {
        const cacheResult = cache.readQuery<GetVersionConfStatus>({
          query: GetRuntimeAndVersionQuery,
          variables: {
            runtimeId: runtimeId,
          }
        });

        const versions = cacheResult?.versions || [];
        const newVersion = updateResult.data
          .createVersion as GetVersionConfStatus_versions;

        cache.writeQuery({
          query: GetRuntimeAndVersionQuery,
          variables: { runtimeId },
          data: { versions: versions.concat([newVersion]) }
        });
      }
    },
    onCompleted
  });

  const [validationErrs, setValidationErrs] = useState<string[]>([]);

  function onCompleted(updatedData: CreateVersion) {
    const versionCreatedId = updatedData.createVersion.id;
    console.log(`${versionCreatedId} version created`);

    history.push(buildRoute.version(ROUTE.VERSION, runtimeId, updatedData.createVersion.name), runtimeId);
  }

  useEffect(() => {
    if (error) {
      setError('addVersionFile', 'apolloError', error.toString());

      const err = error.graphQLErrors[0];
      if (err.extensions?.code === 'krt_validation_error') {
        setValidationErrs(err.extensions?.details);
      }
    } else {
      clearError('addVersionFile');
    }
  }, [error, clearError, setError]);

  function onChange() {
    clearError('addVersionFile');
    setValidationErrs([]);
  }
  function onCancelClick() {
    history.goBack();
  }

  function onSubmit(formData: any) {
    addVersion(mutationPayloadHelper({ file: formData.addVersionFile[0], runtimeId }));
  }

  function getValidationErrors() {
    return (
      <div className={styles.errorBox}>
        The krt.yml file contains the following validation errors:
        <ul>
          {validationErrs.map((err: string) => <li>{err}</li>)}
        </ul>
      </div>
    );
  }

  return (
    <div className={styles.bg}>
      <div className={styles.grid}>
        <div className={styles.container}>
          <h1>Add Version</h1>
          <p className={styles.subtitle} />

          {validationErrs.length > 0 && getValidationErrors()}

          <div className={styles.content} data-testid='uploadVersion'>
            <form data-testid='fileUpload'>
              <FileUpload
                name="addVersionFile"
                inputRef={register({
                  required: 'This field is mandatory',
                  validate: value =>
                    (value[0] && value[0].name.includes('.krt')) ||
                    'Must be a .krt file'
                })}
                label="upload version file (KRT file)"
                placeholder=".krt"
                error={get(errors.addVersionFile, 'message', '')}
                onChange={onChange}
                autofocus
              />
              {loading && (
                <div className={styles.spinner}>
                  <SpinnerLinear />
                </div>
              )}
              <div className={styles.buttons} data-testid='createVersion'>
                <Button
                  primary
                  disabled={!!errors.addVersionFile || loading}
                  label={'CREATE'}
                  onClick={handleSubmit(onSubmit)}
                  className={styles.buttonSave}
                  tabIndex={0}
                />
                <Button label="CANCEL" onClick={onCancelClick} tabIndex={0} />
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default AddVersion;
