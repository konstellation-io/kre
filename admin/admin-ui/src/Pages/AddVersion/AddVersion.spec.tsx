import AddVersion from './AddVersion';
import React from 'react';
import { shallow } from 'enzyme';

jest.mock('@apollo/client', () => ({
  useMutation: jest.fn(() => [jest.fn(), { loading: false, error: '' }]),
  useApolloClient: jest.fn()
}));
jest.mock('react-router', () => ({
  useHistory: jest.fn(() => ({})),
  useParams: jest.fn(() => ({ runtimeId: 'runtimeId' }))
}));

describe('AddVersion', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallow(<AddVersion />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
