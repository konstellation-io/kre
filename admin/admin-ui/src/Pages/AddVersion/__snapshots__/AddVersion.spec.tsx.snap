// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddVersion matches snapshot 1`] = `
<div
  className="bg"
>
  <div
    className="grid"
  >
    <div
      className="container"
    >
      <h1>
        Add Version
      </h1>
      <p
        className="subtitle"
      />
      <div
        className="content"
      >
        <form>
          <Component
            autofocus={true}
            error=""
            inputRef={[Function]}
            label="upload version file (KRT file)"
            name="addVersionFile"
            onChange={[Function]}
            placeholder=".krt"
          />
          <div
            className="buttons"
          >
            <A
              className="buttonSave"
              disabled={false}
              label="CREATE"
              onClick={[Function]}
              primary={true}
              tabIndex={0}
            />
            <A
              label="CANCEL"
              onClick={[Function]}
              tabIndex={0}
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
